BrainPortal/app/controllers/data_providers_controller.rb

```diff
@@ -399,263 +399,329 @@ def browse
+    ensure_browsable(@provider, 'delete', proc do |html, data, status|
       respond_to do |format|
-        format.html { redirect_to :action => :index }
-        format.xml  { render :xml   => { :error  =>  flash[:error] }, :status  => :forbidden }
-        format.json { render :json  => { :error  =>  flash[:error] }, :status  => :forbidden }
+        format.html(&html)
+        format.xml  { render :xml  => data, :status => status }
+        format.json { render :json => data, :status => status }
+      end
+      return
+    end)
+
+    # Is there an automatic copy/move operation to do afterwards?
+    post_action = :copy if params[:auto_do] == "COPY"
+    post_action = :move if params[:auto_do] == "MOVE"
+    target_dp   = DataProvider.find_accessible_by_user(params[:other_data_provider_id], current_user) rescue nil
+    if post_action && ! target_dp
+      flash[:error] = "Missing destination data provider for copy or move."
+      respond_to do |format|
+        format.html { redirect_to :action => :browse }
+        format.xml  { render :xml  => { :error => flash[:error] }, :status => :unprocessable_entity }
+        format.json { render :json => { :error => flash[:error] }, :status => :unprocessable_entity }
       end
       return
     end
 
-    basenames = params[:basenames] || []
-    filetypes = params[:filetypes] || []
-    basenames = [basenames] unless basenames.is_a? Array
-    filetypes = [filetypes] unless filetypes.is_a? Array
+    # Provided file types for the file(s) to register. If a file is present
+    # in 'basenames' (to be registered) but *not* in 'filetypes', a default
+    # of 'SingleFile' is used.
+    filetypes = Array(params[:filetypes])
+      .map { |v| [$2, $1] if v.match(/^(\w+)-(\S+)$/) }
+      .compact
+      .to_h
+
+    # Known userfile types, used to validate the values extracted above
+    valid_types = Userfile.descendants
+      .map(&:name)
+
+    # The new file(s)'s default project is the currently active project, if
+    # available.
+    group_id = current_project.try(:id) || current_user.own_group.id
+
+    # Unless one was specified explicitly via :other_group_id
+    group_id = params[:other_group_id].to_i unless
+      params[:other_group_id].blank?
+
+    # Fallback to the user's own project if the one selected above is invalid
+    # (the everyone project, under which no file is ever registered, or a
+    # project the user doesn't have access to).
+    group_id = current_user.own_group.id if (
+      group_id == Group.everyone.id ||
+      ! current_user.available_groups.map(&:id).include?(group_id)
+    )
+
+    # Register the given userfiles in background.
+    userfiles = userfiles_from_basenames(@provider, @as_user, params[:basenames])
+    userfiles_count = userfiles.count # Avoids a cute race condition
+
+    registered, already_registered = [], []
+    succeeded, failed = [], {}
+
+    CBRAIN.spawn_with_active_records_if(
+      api_request,
```

Comment:
Hum, do I understand properly, we spawn ONLY if it's an API request? No when it's a web-based, interactive session?